src/main/scala/com/sksamuel/scapegoat/io/GitlabCodeQualityReportWriter.scala

```diff
@@ -0,0 +1,114 @@
+package com.sksamuel.scapegoat.io
+
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+
+import com.sksamuel.scapegoat.{Feedback, Levels, Warning}
+
+/**
+ * Supports GitLab Code Quality report format.
+ *
+ * https://docs.gitlab.com/ee/ci/testing/code_quality.html#implement-a-custom-tool
+ */
+object GitlabCodeQualityReportWriter extends ReportWriter {
+
+  override protected def fileName: String = "scapegoat-gitlab.json"
+
+  override protected def generate(feedback: Feedback): String =
+    toCodeQualityElements(feedback.warningsWithMinimalLevel, sys.env.get("CI_PROJECT_DIR"))
+      .map(_.toJsonArrayElement)
+      .mkString("[", ",", "]")
+
+  def toCodeQualityElements(
+    warnings: Seq[Warning],
+    gitlabBuildDir: Option[String]
+  ): Seq[CodeQualityReportElement] = warnings.map { warning =>
+    // Stable hash for the same warning.
+    // Avoids moving code blocks around from causing "new" detecions.
+    val fingerprintRaw = warning.sourceFileNormalized + warning.snippet.getOrElse(warning.line.toString)
+    val fingerprint = MessageDigest
+      .getInstance("MD5")
```

Comment:
Maybe extract the `MessageDigest` to only have one instance instead of initialising one for every warning?

Reply:
As MessageDigest is stateful and not thread safe, I've moved it to the outer loop so it at least re-uses the instance for an entire report.